examples/kalman_filter.py

```diff
@@ -0,0 +1,374 @@
+
+    def compute_gain(self, observation):
+        """Compute the Kalman gain given the observation model."""
+        N = self.model.get_measurement_noise_cov(
+            self.state, self.measurement_noise)
+        obs_jac = self.model.observation_jacobian(self.state, observation)
+        innovation_cov = obs_jac.dot(self.covariance).dot(obs_jac.T) + N
+        return self.covariance.dot(obs_jac.T).dot(
+            gs.linalg.inv(innovation_cov))
+
+    def update(self, observation):
+        """Update the current estimate given an observation."""
+        innovation = self.model.innovation(self.state, observation)
+        gain = self.compute_gain(observation)
+        obs_jac = self.model.observation_jacobian(self.state, observation)
+        self.covariance = (gs.eye(self.model.dim) - gain.dot(obs_jac)).dot(
+            self.covariance)
+        self.state = self.model.group.exp(gain.dot(innovation), self.state)
+        if gs.ndim(self.state) > 1:
+            return gs.squeeze(self.state)
+
+
+def main():
+    """Carry out two examples of state estimation on groups.
+
+     Both examples are localization problems, where only a part of the system
+     is observed. The first one is a linear system, while the second one is
+     non-linear.
+     """
+    def create_data(kalman, true_init, true_inputs, obs_freq):
+        """Create data for a specific example."""
+        true_traj = [1 * true_init]
+        for incr in true_inputs:
+            true_traj.append(kalman.model.propagate(true_traj[-1], incr))
+        true_obs = [
+            kalman.model.observation_model(pose)
+            for pose in true_traj[obs_freq::obs_freq]]
+
+        observations = [
+            np.random.multivariate_normal(obs, kalman.measurement_noise)
+            for obs in true_obs]
+
+        inputs = [gs.concatenate(
+            (incr[:1], np.random.multivariate_normal(
+                incr[1:], kalman.process_noise))) for incr in true_inputs]
+
+        return gs.array(true_traj), inputs, gs.array(observations)
+
+    def estimation(observer, initial_state, inputs, obs, obs_freq):
+        """Carry out the state estimation for a specific system."""
+        observer.state = 1 * initial_state
+
+        traj = [1 * observer.state]
+        uncertainty = [1 * gs.diagonal(observer.covariance)]
+        for i in range(n_traj):
+            observer.propagate(inputs[i])
+            if i > 0 and i % obs_freq == obs_freq - 1:
+                observer.update(obs[(i // obs_freq)])
+            traj.append(1 * observer.state)
+            uncertainty.append(1 * gs.diagonal(observer.covariance))
+        traj = gs.array(traj)
+        uncertainty = gs.array(uncertainty)
+        three_sigmas = 3 * gs.sqrt(uncertainty)
+
+        return traj, three_sigmas
+
+    np.random.seed(12345)
+    model = LocalizationLinear()
+    kalman = KalmanFilter(model)
+    n_traj = 1000
+    obs_freq = 50
+    dt = 0.1
+    P0 = gs.array([10., 1.])
```

Comment:
caps are for global variables IIRC. @ninamiolane can you confirm?

Reply:
Yes! The only exception that we have made is `X`, the design matrix, to follow scikit learn notations.